symengine/dict.h

```diff
@@ -49,20 +49,30 @@ typedef std::map<unsigned, integer_class> map_uint_mpz;
 typedef std::map<unsigned, rational_class> map_uint_mpq;
 typedef std::map<int, Expression> map_int_Expr;
 typedef std::vector<integer_class> vec_integer_class;
-//! Part of umap_vec_mpz:
-typedef struct {
-    inline std::size_t operator()(const vec_int &k) const
+typedef std::vector<unsigned int> vec_uint;
+typedef std::unordered_map<RCP<const Basic>, unsigned int, RCPBasicHash,
+                           RCPBasicKeyEq> umap_basic_uint;
+
+template <typename T>
+class vec_hash
+{
+public:
+    std::size_t operator()(const T &v) const
     {
         std::size_t h = 0;
-        for (const auto &p : k) {
-            h = (h << 4) + p;
+        for (auto i : v) {
+            h ^= i + 0x9e3779b + (h << 6) + (h >> 2);
```

Comment:
It's better to use `hash_combine` to combine hashes instead of writing algorithms like this. Only place where we don't use `hash_combine` is `Add`, but `Add` is special than others. If `hash_combine` isn't good, we can change that later.